walkeeper/src/callmemaybe.rs

```diff
@@ -0,0 +1,131 @@
+//!
+//!  Callmemaybe thread is responsible for periodically requesting
+//!  pageserver to initiate wal streaming.
+//!
+use crate::receive_wal::WalStreamInfo;
+use crate::SafeKeeperConf;
+use anyhow::Result;
+use crossbeam_channel::{Receiver, RecvTimeoutError};
+use log::*;
+use std::sync::Arc;
+use std::time::Duration;
+use tokio_postgres::NoTls;
+use zenith_utils::zid::{ZTenantId, ZTimelineId};
+
+async fn request_callback(
+    ps_addr: String,
+    listen_pg_addr: String,
+    pageserver_auth_token: Option<String>,
+    timelineid: ZTimelineId,
+    tenantid: ZTenantId,
+) -> Result<()> {
+    // Connect to the pageserver.
+    let ps_addr_split: Vec<&str> = ps_addr.split(':').collect();
+    let mut config = tokio_postgres::Config::new();
+
+    config
+        .host(ps_addr_split[0])
+        .port(ps_addr_split[1].parse().unwrap());
+
+    if let Some(password) = pageserver_auth_token {
+        config.password(password);
+    }
+
+    debug!("request_callback Connecting to pageserver {}", ps_addr);
+    let (client, connection) = config.connect(NoTls).await?;
+
+    tokio::spawn(async move {
+        if let Err(e) = connection.await {
+            eprintln!("connection error: {}", e);
+        }
+    });
+
+    // Send callmemaybe request
+    let pg_addr_split: Vec<&str> = listen_pg_addr.split(':').collect();
+    let (host, port) = (pg_addr_split[0], pg_addr_split[1]);
+
+    let callme = format!(
+        "callmemaybe {} {} host={} port={} options='-c ztimelineid={} ztenantid={}'",
+        tenantid, timelineid, host, port, timelineid, tenantid
+    );
+
+    let _ = client.simple_query(&callme).await?;
+
+    Ok(())
+}
+
+pub fn thread_main(conf: SafeKeeperConf, rx: Receiver<Arc<WalStreamInfo>>) -> Result<()> {
+    let runtime = tokio::runtime::Builder::new_current_thread()
+        .enable_all()
+        .build()?;
+
+    let default_timeout = Duration::from_secs(5);
+    let timeout = conf.recall_period.unwrap_or(default_timeout);
+    let mut addresses = Vec::new();
+
+    loop {
+        match rx.recv_timeout(timeout) {
+            Ok(streaminfo) => {
+                info!("callmemaybe. thread_main. request_callback for timelineid={} tenantid={} refcount {}",
+                streaminfo.timelineid, streaminfo.tenantid, Arc::strong_count(&streaminfo));
```

Comment:
@LizardWizzard has proposed a brilliant idea: we could implement "unsubscribe" as a guard, i.e. a struct with a destructor. The destructor could send an "unsubscribe" event containing `TentantId` and whatnots via a [`Sender`](https://docs.rs/tokio/1.14.0/tokio/sync/mpsc/struct.Sender.html).

It's also worth mentioning that tokio's mpsc supports [blocking send](https://docs.rs/tokio/1.14.0/tokio/sync/mpsc/struct.Sender.html#method.blocking_send).